src/shader/shaderLightSampling.js

```diff
@@ -27,20 +67,50 @@ LightSampleRec lightsClosestHit( sampler2D lights, uint lightCount, vec3 rayOrig
 			continue;
 		}
 
-		u *= 1.0 / dot(u, u);
-		v *= 1.0 / dot(v, v);
+		u *= 1.0 / dot( u, u );
+		v *= 1.0 / dot( v, v );
 
 		float dist;
-		if ( intersectsRectangle( light.position, normal, u, v, rayOrigin, rayDirection, dist ) ) {
 
-			if ( dist < lightSampleRec.dist || !lightSampleRec.hit ) {
+		if( light.type == 0 ) {
+
+			if ( intersectsRectangle( light.position, normal, u, v, rayOrigin, rayDirection, dist ) ) {
+
+				if ( dist < lightSampleRec.dist || !lightSampleRec.hit ) {
+
+					lightSampleRec.hit = true;
+					lightSampleRec.dist = dist;
+					float cosTheta = dot( rayDirection, normal );
+					lightSampleRec.pdf = ( dist * dist ) / ( light.area * cosTheta );
+					lightSampleRec.emission = light.color * light.intensity;
+					lightSampleRec.direction = rayDirection;
+
+				}
+
+			}
+
+		} else {
+
+			SpotLight spotLight = readSpotLightInfo( lights, l );
+
+			float angle = acos( spotLight.coneCos );
+			float angleTan = tan( angle );
+			float startDistance = spotLight.radius / angleTan;
+
+			vec3 lightPosition = light.position - normal * startDistance;
+			if ( spotLight.lampIntensityScale > 0.0 && spotLight.radius > 0.0 && intersectsCircle( lightPosition, normal, u, v, spotLight.radius * 2.0, rayOrigin, rayDirection, dist ) ) {
+
+				if ( dist < lightSampleRec.dist || !lightSampleRec.hit ) {
+
+					lightSampleRec.hit = true;
+					lightSampleRec.dist = dist;
+					float cosTheta = dot( rayDirection, normal );
+					lightSampleRec.pdf = ( dist * dist ) / ( light.area * cosTheta );
+					lightSampleRec.emission = light.color * light.intensity * spotLight.lampIntensityScale;
```

Comment:
I think it's okay if we skip the `lampIntensityScale` factor for now - with this disabled the lighting does not work in the non MIS case or the primary path tracing. I think we'd want a more general approach that works for disabling both area lights and spot light surfaces -- I'll make an issue for it.

Reply:
I didn't like this concept of "lamp" here but I put it in to be more physically accurate, this probably goes against how most scenes are built hence why I opted to make this feature an optional / convenience thing. In actuality people probably would use a textured and emissive surface, say for a theatre lamp.

Reply:
I think one of the disconnects here is that `lampIntensityScale` is actually attenuating the sample intensity, as well, whereas to get the effect you're talking about you really just want it to scale the intensity if the first ray (or maybe specular ray) hits it directly if I'm understanding right.

See the difference in effect here with Non MIS rendering:

| 1.0 lampIntensityScale | 0.1 lampIntensityScale  |
|---|---|
| ![image](https://user-images.githubusercontent.com/734200/175754024-a3ee65ff-49b3-405a-9ba8-c24cf378f54b.png) | ![image](https://user-images.githubusercontent.com/734200/175754034-3d8f8cc4-057b-40aa-9def-7bde6b584b8a.png) |

I agree with you that there's value in being able to hide the light surface but I think it should probably just be a boolean and work for all lights so I think it's okay if that comes later. I added #193 to track it.